src/Tasks/Microsoft.NET.Build.Tasks/PrepareForReadyToRunCompilation.cs

```diff
@@ -248,16 +214,131 @@ protected override void ExecuteCore()
+                r2rFilesPublishList.Add(compositeR2RFileToPublish);
+            }
         }
 
-        private enum Eligibility
+        private struct Eligibility
         {
-            None,
-            ReferenceOnly,
-            CompileAndReference
+            [Flags]
+            private enum EligibilityEnum
+            {
+                None = 0,
+                Reference = 1,
+                HideReferenceFromComposite = 2,
+                CompileSeparately = 4,
+                CompileIntoCompositeImage = 8,
+            }
+
+            private readonly EligibilityEnum _flags;
+
+            public static Eligibility None => new Eligibility(EligibilityEnum.None);
+
+            public bool NoEligibility => _flags == EligibilityEnum.None;
+            public bool IsReference => (_flags & EligibilityEnum.Reference) == EligibilityEnum.Reference;
+            public bool ReferenceHiddenFromCompositeBuild => (_flags & EligibilityEnum.HideReferenceFromComposite) == EligibilityEnum.HideReferenceFromComposite;
+            public bool CompileIntoCompositeImage => (_flags & EligibilityEnum.CompileIntoCompositeImage) == EligibilityEnum.CompileIntoCompositeImage;
+            public bool CompileSeparately => (_flags & EligibilityEnum.CompileSeparately) == EligibilityEnum.CompileSeparately;
+            public bool Compile => CompileIntoCompositeImage || CompileSeparately;
+
+            private Eligibility(EligibilityEnum flags)
+            {
+                _flags = flags;
+            }
+
+            public static Eligibility CreateReferenceEligibility(bool hideFromCompositeBuilds)
+            {
+                if (hideFromCompositeBuilds)
+                    return new Eligibility(EligibilityEnum.Reference | EligibilityEnum.HideReferenceFromComposite);
+                else
+                    return new Eligibility(EligibilityEnum.Reference);
+            }
+
+            public static Eligibility CreateCompileEligibility(bool doNotBuildIntoComposite)
+            {
+                if (doNotBuildIntoComposite)
+                    return new Eligibility(EligibilityEnum.Reference | EligibilityEnum.HideReferenceFromComposite | EligibilityEnum.CompileSeparately);
+                else
+                    return new Eligibility(EligibilityEnum.Reference | EligibilityEnum.CompileIntoCompositeImage);
+            }
         };
 
-        private static Eligibility GetInputFileEligibility(ITaskItem file, HashSet<string> exclusionSet)
+        private static bool IsNonCompositeReadyToRunImage(PEReader peReader)
+        {
+            if (peReader.PEHeaders == null)
+                return false;
+
+            if (peReader.PEHeaders.CorHeader == null)
+                return false;
+
+            if ((peReader.PEHeaders.CorHeader.Flags & CorFlags.ILLibrary) == 0)
+            {
+                // This is likely a composite image, but those can't be re-r2r'd
```

Comment:
I don't see how we could arrive here for a composite image as it has no CorHeader, can you please clarify?

Reply:
Actually, it turns out that composite images DO have a CorHeader. In fact, this code is a copy from r2rdump, but with the special handling to find the `RTR_HEADER` export removed.